### PR TITLE
Fpu divider based on Radix-4 SRT divider.

### PR DESCRIPTION
**Pull request by Xusine1131**

### CODE REVIEW

**Comment by taylor-bsg** [comment]
Wow! This is neat! @Xusine1131  

@tommydcjung please review carefully — you are the main expert in the group in this.

**Comment by tommydcjung** [comment]
some comments made. I will take closer look at the code soon

File: testing/bsg_fpu/div_srt_32/bsg_test_dpi.c [hunk 1]
```diff
@@ -0,0 +1,21 @@
+#include<svdpi.h>
```
[the remark this answers is not in the document]
Comment (Xusine1131) [reply testing/bsg_fpu/div_srt_32/bsg_test_dpi.c]: It seems that you put the trace and result into a ROM file, and then using a C file to make comparison? If so, I think SystemVerilog DPI is very suitable for this situation.
I'm not sure whether DPI is banned in basejump team but it can really make fp test convenient.

[the remark this answers is not in the document]
Comment (dpetrisko) [reply testing/bsg_fpu/div_srt_32/bsg_test_dpi.c]: DPI is fine in general.  I think Tommy's concern is that the test shouldn't be tied to x86's implementation of fp.  But if they're both IEEE compliant, probably shouldn't cause an issue?

For what it's worth, http://www.jhauser.us/arithmetic/TestFloat.html is the gold standard for testing IEEE floating point

[the remark this answers is not in the document]
Comment (Xusine1131) [reply testing/bsg_fpu/div_srt_32/bsg_test_dpi.c]: @dpetrisko Thanks for your suggestion, I'll check it out and rewrite my test file!

[the remark this answers is not in the document]
Comment (tommydcjung) [reply testing/bsg_fpu/div_srt_32/bsg_test_dpi.c]: you are free to use whatever technology you prefer. my point is that all the corner cases and their exact expected outcome should be documented like in README (e.g. what happens when the input is denormal,  or what should be outputted when there is invalid operation, etc), so that when we run some serious randomized tests, we know exactly what to expect. Besides those corner cases, there should be no difference in computed result.

File: testing/bsg_fpu/div_srt_32/test_bsg_float32.v [hunk 2]
```diff
@@ -0,0 +1,136 @@
...
+    else if(fp32_underflow) begin
+      $display("Underflow condition occurs.");
+    end
+    else if(fp32_invalid) begin
+      $display("Invalid condition occurs.");
+    end
+    else begin
+      res_module = float32Encode(fp32_result_o);
+      res_vcs = float32Encode(fp32_dividend_i)/float32Encode(fp32_divisor_i);
+      if(res_module - res_vcs > 0.000001 ||res_module - res_vcs  < -0.000001) begin
```
Comment (tommydcjung) [inline testing/bsg_fpu/div_srt_32/test_bsg_float32.v]: if there is even one bit difference, that should count as an error.

File: bsg_fpu/bsg_fpu_div.v [hunk 3]
```diff
@@ -0,0 +1,311 @@
...
+wire [divider_width_lp:0] round_input = result_mantissa_r[divider_width_lp] ? result_mantissa_r : {result_mantissa_r[divider_width_lp-1:0], result_mantissa_r[0]}; // sticky bit
+
+bsg_fpu_round #(
+  .width_i_p(divider_width_lp+1)
+  ,.width_o_p(m_p+1)
+) rounder (
+  .type_i(bsg_fpu_pkg::eRtne)
+  ,.mantissa_i(round_input)
+  ,.sign_i(result_sign_r)
+  ,.mantissa_o(round_out)
```
Comment (tommydcjung) [inline bsg_fpu/bsg_fpu_div.v]: what happens when the mantissa overflows after rounding? shouldn't that carry over to exponents?

**Comment by taylor-bsg** [comment]
+1 on Hauser’s Test infrastructure.

Intel floating point is 80-bit internally, so it can get better results
when you do a series of floating point ops, but if you are just doing
divide the yeah, I wouldn’t see why we should have divergence except for
different in rounding mode support.

File: testing/bsg_fpu/div_srt_32/bsg_test_dpi.c (hunk 1 above)
[the remark this answers is not in the document]
Comment (Xusine1131) [reply testing/bsg_fpu/div_srt_32/bsg_test_dpi.c]: @tommydcjung I get the point, thanks!

**Comment by Xusine1131** [comment]
I think these difference may come from SRT divider. I have checked several divergence and they have something with incorrect quotient selection. 
Let me fix these bugs at first. I'm very sorry to bother you @taylor-bsg @dpetrisko @tommydcjung 
@tommydcjung Thanks for your feedback anyway!
